middle_end/flambda2/simplify/simplify_let_cont_expr.ml

```diff
@@ -1274,14 +1505,74 @@ let after_downwards_traversal_of_body ~simplify_expr
     in
     let common_denv = DA.denv dacc in
     assert (not is_exn_handler);
-    simplify_recursive_handlers ~rebuild_body ~invariant_params ~invariant_epa
-      ~down_to_up ~continuation_handlers ~simplify_expr
+    simplify_recursive_handlers ~rebuild_body ~lifted_params ~invariant_params
+      ~invariant_epa ~continuation_handlers ~simplify_expr
       ~consts_lifted_during_body ~all_conts_set ~common_denv
       body_continuation_uses_env used_handlers_in_body Continuation.Set.empty
-      Continuation.Map.empty dacc
+      Continuation.Map.empty dacc (fun dacc data ->
+        let dacc = DA.with_are_lifting_conts dacc previous_are_lifting_conts in
+        k dacc data)
 
-let simplify_let_cont0 ~simplify_expr dacc (data : simplify_let_cont_data)
-    ~down_to_up =
+and after_downwards_traversal_of_body ~simplify_expr ~down_to_up
+    (data : after_downwards_traversal_of_body_data) dacc ~rebuild:rebuild_body =
+  (* At this point, we have done the downwards traversal of the body, and we
+     have two situations wrt to continuation lifting. *)
+  let denv_for_join = data.denv_for_join in
+  match DA.are_lifting_conts dacc with
+  | Lifting_out_of { continuation = _ } ->
+    (* In this case, we have decided to lift the continuation being bound out of
+       its defining handler. Therefore we save the non-simplified version of the
+       continuation in the dacc, so that we can simplify during the `down_to_up`
+       later. *)
+    let params_to_lift =
+      DE.variables_defined_in_current_continuation (DA.denv dacc)
+    in
+    let handlers =
+      Lifted_cont.add_params_to_lift data.handlers params_to_lift
+    in
+    let dacc = DA.add_lifted_continuation data.denv_for_join handlers dacc in
+    (* Restore lifted constants in dacc *)
+    let dacc =
+      DA.add_to_lifted_constant_accumulator dacc data.prior_lifted_constants
+    in
+    let data : after_downwards_traversal_of_body_and_handlers_data =
+      Lifted_out { rebuild_body }
+    in
+    after_downwards_traversal_of_body_and_handlers ~simplify_expr ~denv_for_join
+      data ~down_to_up dacc
+  | Not_lifting | Analyzing _ ->
+    simplify_handlers ~simplify_expr data dacc ~rebuild_body (fun dacc data ->
+        after_downwards_traversal_of_body_and_handlers ~simplify_expr
+          ~denv_for_join
+          (Rebuild data : after_downwards_traversal_of_body_and_handlers_data)
+          ~down_to_up dacc)
```

Comment:
For readability purpose, I would rather have simplify handler build his own continuation rather than building it here. It just need `denv_for_join` and `down_to_up` as additionnal arguments